### PR TITLE
Support block-wise fp8 quant

### PR DESCRIPTION
**Pull request by mengniwang95**
## Description

Support block-wise fp8 quant

https://github.com/intel/auto-round/issues/959

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Performance improvement
- [ ] Code refactoring
- [ ] Other (please specify):

### CODE REVIEW

File: auto_round/compressors/config.py [hunk 1]
```diff
@@ -49,6 +49,7 @@ def __init__(
         group_size: int = None,
         sym: bool = None,
         data_type: str = None,
+        weight_block_size: list = None,
```
Comment (wenhuach21) [inline auto_round/compressors/config.py]: how about merging this param to group_size

File: auto_round/schemes.py [hunk 2]
```diff
@@ -244,6 +245,18 @@ def is_preset_scheme(name: str) -> bool:
     }
 )
 
+FP8_BLOCK = QuantizationScheme.from_dict(
```
Comment (wenhuach21) [inline auto_round/schemes.py:247]: what's the scheme name in llmc, we may use the same name

Comment (mengniwang95) [reply auto_round/schemes.py:247]: it also uses FP8_BLOCK

File: auto_round/data_type/utils.py [hunk 3]
```diff
@@ -44,6 +44,13 @@ def reshape_pad_tensor_by_group_size(data: torch.Tensor, group_size: int, val: f
     """
     orig_shape = data.shape
     pad_len = 0
+    if isinstance(group_size, list):
+        assert len(group_size) == 2, f"Only support 2D weight_block_size, but get {len(group_size)}"
+        M, N = group_size
+        pad_len_m = (orig_shape[0] + M - 1) // M * M - orig_shape[0]
+        pad_len_n = (orig_shape[1] + M - 1) // N * N - orig_shape[1]
```
Comment (yiliu30) [inline auto_round/data_type/utils.py]:
```suggestion
        pad_len_n = (orig_shape[1] + N - 1) // N * N - orig_shape[1]
```

Comment (yiliu30) [reply auto_round/data_type/utils.py]: nit: better use a ceil func

File: auto_round/compressors/base.py [hunk 4]
```diff
@@ -856,6 +865,10 @@ def _check_configs(self) -> None:
         if is_nv_fp(self.data_type) and (self.group_size != 16):
             logger.warning("dtype nv_fp should only support group_size of 16 in real deployment")
 
+        if self.weight_block_size is not None and not is_block_wfp8(self):
+            logger.error("only support block-wise quantization for fp8 weight quantization.")
+            exit(-1)
```
Comment (yiliu30) [inline auto_round/compressors/base.py]: It would be better to raise a NotImplementedError instead of exiting silently.

File: auto_round/data_type/utils.py [hunk 5]
```diff
@@ -44,6 +44,13 @@ def reshape_pad_tensor_by_group_size(data: torch.Tensor, group_size: int, val: f
     """
     orig_shape = data.shape
     pad_len = 0
+    if isinstance(group_size, list):
```
Comment (yiliu30) [inline auto_round/data_type/utils.py]: Quick question: if we apply padding, will the final quantized weight (on disk) have a different shape from the original one?

File: test/test_cpu/quantization/test_block_fp.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        )
+        quantized_model_path = self.save_dir
+        compressed_model, _ = autoround.quantize_and_save(output_dir=quantized_model_path, format="auto_round")
+        tmp_layer = compressed_model.model.decoder.layers[1].self_attn.q_proj
+        assert (
+            hasattr(tmp_layer, "weight_scale_inv")
+            and tmp_layer.weight.dtype is torch.float8_e4m3fn
+            and list(tmp_layer.weight_scale_inv.shape) == [6, 6]
+        ), "Illegal FP8_BLOCK packing name or data_type or shape"
+        shutil.rmtree(quantized_model_path, ignore_errors=True)
```
Comment (yiliu30) [inline test/test_cpu/quantization/test_block_fp.py:152]: Please add some asserts here.

File: auto_round/data_type/utils.py (hunk 3 above)
Comment (mengniwang95) [reply auto_round/data_type/utils.py]: updated

File: auto_round/compressors/base.py (hunk 4 above)
Comment (mengniwang95) [reply auto_round/compressors/base.py]: updated

File: auto_round/data_type/utils.py (hunk 5 above)
Comment (mengniwang95) [reply auto_round/data_type/utils.py]: no, the shape is the original one

File: test/test_cpu/quantization/test_block_fp.py (hunk 6 above)
Comment (mengniwang95) [reply test/test_cpu/quantization/test_block_fp.py:152]: add more tests

File: auto_round/wrapper.py [hunk 7]
```diff
@@ -57,21 +62,28 @@ def reshape_and_pad_tensor(v, group_size=-1):
...
-    if group_size == -1 or v.shape[1] < group_size:
-        return v
-    if v.shape[1] % group_size == 0:
-        v = v.reshape(-1, group_size)
+    if isinstance(group_size, list):
+        assert len(group_size) == 2, f"Only support 2D group_size, but get {group_size}"
+        M, N = group_size
+        pad_len_m = ceil(v.shape[0] / M) * M - v.shape[0]
+        pad_len_n = ceil(v.shape[1] / N) * N - v.shape[1]
+        v = torch.nn.functional.pad(v, (pad_len_m, pad_len_n))
```
Comment (yiliu30) [inline auto_round/wrapper.py]:
```suggestion
        v = torch.nn.functional.pad(v, (0, pad_len_n, 0, pad_len_m))
```
We need to pad the columns and rows, right? The current implementation is incorrect. 
Please double-check the pad function and add some UTs for this part. 
https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
```bash
  v = torch.nn.functional.pad(v, (0, pad_len_n, 0, pad_len_m))
  #                               │  │          │  └─ pad rows at bottom
  #                               │  │          └──── no padding at top
  #                               │  └─────────────── pad cols at right
  #                               └────────────────── no padding at left
```

File: auto_round/__main__.py [hunk 8]
```diff
@@ -285,7 +285,12 @@ def __init__(self, *args, **kwargs):
 
         scheme = self.add_argument_group("Scheme Arguments")
         scheme.add_argument("--bits", default=None, type=int, help="Number of bits for weight quantization. ")
-        scheme.add_argument("--group_size", default=None, type=int, help="Group size for weight quantization.")
+        scheme.add_argument(
+            "--group_size",
+            default=None,
+            type=lambda s: int(s) if s.isdigit() else [int(x.strip()) for x in s.split(",")],
```
Comment (yiliu30) [inline auto_round/__main__.py]: Please help double-check the impact on `--group_size`.
`--group_size -1` -> [-1] ?

File: auto_round/compressors/base.py [hunk 9]
```diff
@@ -827,10 +829,12 @@ def _check_configs(self) -> None:
             raise ValueError("`bits` must be positive")
         if self.act_bits <= 0:
             raise ValueError("`act_bits` must be positive")
-        if not (self.group_size == -1 or self.group_size >= 0):
+        if not isinstance(self.group_size, list) and not (self.group_size == -1 or self.group_size >= 0):
             raise ValueError("`group_size` must be -1 (per channel) or 0 (per-tensor) or a positive integer")
-        if not (self.act_group_size == -1 or self.act_group_size >= 0):
+        if not isinstance(self.group_size, list) and not (self.act_group_size == -1 or self.act_group_size >= 0):
```
Comment (yiliu30) [inline auto_round/compressors/base.py]: nit: It would be better to define a function for that check.

File: test/test_cpu/quantization/test_block_fp.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+            seqlen=2,
+        )
+        quantized_model_path = self.save_dir
+        compressed_model, _ = autoround.quantize_and_save(output_dir=quantized_model_path, format="llm_compressor")
+        tmp_layer = compressed_model.model.decoder.layers[1].self_attn.q_proj
+        assert hasattr(tmp_layer, "weight_scale")
+        assert tmp_layer.weight.dtype is torch.float8_e4m3fn
+        assert list(tmp_layer.weight_scale.shape) == [6, 6]
+        assert compressed_model.config.quantization_config["quant_method"] == "compressed-tensors"
+        shutil.rmtree(quantized_model_path, ignore_errors=True)
```
Comment (yiliu30) [inline test/test_cpu/quantization/test_block_fp.py:152]: Please add some UTs to load the quantized model and do smoke inference.

File: auto_round/__main__.py (hunk 8 above)
Comment (mengniwang95) [reply auto_round/__main__.py]: `--group_size -1` -> -1

File: auto_round/wrapper.py (hunk 7 above)
Comment (mengniwang95) [reply auto_round/wrapper.py]:
```python
>>> data = torch.rand(5,4)
>>> padded_data = torch.nn.functional.pad(data, (0, 2, 0, 3))
>>> padded_data.shape
torch.Size([8, 6])
```

Comment (mengniwang95) [reply auto_round/wrapper.py]: pad_len_m is the padded rows number, it pads rows at bottom.
Do you think what is the correct implementation? @yiliu30

File: test/test_cpu/quantization/test_block_fp.py (hunk 10 above)
Comment (mengniwang95) [reply test/test_cpu/quantization/test_block_fp.py:152]: add accuracy checker

File: docs/fp8_block_acc.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+lm_eval --model vllm --model_args pretrained=model_path,add_bos_token=true,tensor_parallel_size=1,gpu_memory_utilization=0.8,max_model_len=8192 --tasks tasks --batch_size 128
+```
+
+## LLaMa-3-8B-Instruct
+
+|Scheme          |arc_challenge|arc_easy|boolq |hellaswag|lambada_openai|openbookqa|piqa  |truthfulqa_mc1|winogrande|gsm8k_llama|mmlu_llama|mmlu_pro_llama|
+|:---------------|:------------|:-------|:-----|:--------|:-------------|:---------|:-----|:-------------|:---------|:----------|:---------|:-------------|
+|BF16            |0.5341       |0.8232  |0.8428|0.5805   |0.7258        |0.3540    |0.7758|0.3709        |0.7151    |0.6111     |0.6737    |0.3778        |
+|FP8_BLOCK RTN   |0.5239       |0.8182  |0.8483|0.5780   |0.7235        |0.3520    |0.7840|0.3647        |0.7040    |0.6171     |0.6681    |0.3738        |
+|FP8_BLOCK Tuning|0.5213       |0.8047  |0.8376|0.5834   |0.7099        |0.3480    |0.7764|0.3684        |0.7245    |0.6224     |0.6654    |0.3690        |
```
Comment (wenhuach21) [inline docs/fp8_block_acc.md]: please give an average, it seems RTN is better. Please recommend users to use RTN when user not using it

File: docs/fp8_block_acc.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
+Below accuracy results are got by lm-eval.
+
+cmd:
+
+```bash
+# model quantization
+auto-round --model model_name_or_path --scheme FP8_BLOCK --iters 0 --format fp8 # RTN
```
Comment (wenhuach21) [inline docs/fp8_block_acc.md:7]: Could AR automatically switch to the FP8 format? If not,  please add a warning that fp8 format is recommend for this scheme

Also, is there a checker to verify whether the current quantization scheme can be exported to the FP8 format?

File: docs/step_by_step_CN.md [hunk 13]
```diff
@@ -146,11 +146,12 @@ AutoRound 支持多种量化配置：
...
 | **auto_awq**    | W4A16、BF16                                                                                                                                                                                                   |
 | **auto_gptq**   | W4A16、W2A16、W3A16、W8A16、W2A16G64、W2A16G32、BF16                                                                                                                                                           |
-| **llm_compressor** | NVFP4、`MXFP4`、`MXFP8`、`FPW8A16`、`FP8_STATIC`                                                                                                                                                              |
+| **llm_compressor** | NVFP4、`MXFP4`、`MXFP8`、`FPW8A16`、`FP8_STATIC`、FP8_STATIC                                                                                                                                                              |
 | **gguf**        | GGUF:Q4_K_M、GGUF:Q2_K_S、GGUF:Q3_K_S、GGUF:Q3_K_M、GGUF:Q3_K_L、GGUF:Q4_K_S、GGUF:Q5_K_S、GGUF:Q5_K_M、GGUF:Q6_K、GGUF:Q4_0、GGUF:Q4_1、GGUF:Q5_0、GGUF:Q5_1、GGUF:Q8_0                                           |
+| **fp8**         | FP8_STATIC  |
```
Comment (wenhuach21) [inline docs/step_by_step_CN.md]: FP8_BLOCK?

File: README.md [hunk 14]
```diff
@@ -36,6 +36,7 @@ See our papers [SignRoundV1](https://arxiv.org/pdf/2309.05516) and [SignRoundV2]
 
 
 ## 🆕 What's New
+* [2026/03] **Block-wise FP8** quantization is available via `--scheme FP8_BLOCK`: [*Accuracy*](./docs/fp8_block_acc.md).
```
Comment (wenhuach21) [inline README.md]: add --format fp8, iters 0, --disable_opt_rtn.
@n1ck-guo  Could we add a cmd auto-round-rtn?

File: auto_round/wrapper.py [hunk 15]
```diff
@@ -37,17 +39,20 @@ def get_scale_shape(weight, group_size):
 
     Args:
       weight (torch.Tensor): The weight tensor of the layer.
-      group_size (int): The size of the groups for quantization.
+      group_size (int or list): The size of the groups for quantization.
```
Comment (wenhuach21) [inline auto_round/wrapper.py]: tuple is better?

File: auto_round/wrapper.py [hunk 16]
```diff
@@ -164,8 +176,36 @@ def _init_tuning_params_and_quant_func(self):
             orig_weight = orig_weight.t()
         weight_reshape = reshape_and_pad_tensor(orig_weight.data, orig_layer.group_size)
         if self.enable_round_tuning:
-            self.weight_min = torch.clamp(weight_reshape.min(1)[0], max=0)
-            self.weight_max = torch.clamp(weight_reshape.max(1)[0], min=0)
+            self.weight_min = (
+                torch.clamp(
+                    weight_reshape.view(
+                        weight_reshape.shape[0] // orig_layer.group_size[0],
```
Comment (wenhuach21) [inline auto_round/wrapper.py]: The implementation looks a bit long. It might be better to wrap it into a function.

File: docs/fp8_block_acc.md (hunk 11 above)
Comment (mengniwang95) [reply docs/fp8_block_acc.md]: updated

File: docs/fp8_block_acc.md (hunk 12 above)
Comment (mengniwang95) [reply docs/fp8_block_acc.md:7]: updated. There are checks for FP8 format

File: docs/step_by_step_CN.md (hunk 13 above)
Comment (mengniwang95) [reply docs/step_by_step_CN.md]: updated

File: README.md (hunk 14 above)
Comment (mengniwang95) [reply README.md]: updated

File: auto_round/wrapper.py (hunk 15 above)
Comment (mengniwang95) [reply auto_round/wrapper.py]: updated

File: auto_round/wrapper.py (hunk 16 above)
Comment (mengniwang95) [reply auto_round/wrapper.py]: code is refined

**Review by wenhuach21 — APPROVED** [review]
